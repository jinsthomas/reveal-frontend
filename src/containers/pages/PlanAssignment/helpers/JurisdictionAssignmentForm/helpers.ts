import moment from 'moment';
import { OpenSRPJurisdiction } from '../../../../../components/TreeWalker/types';
import { PlanDefinition } from '../../../../../configs/settings';
import { Assignment } from '../../../../../store/ducks/opensrp/assignments';

export interface Payload {
  toCreate: Assignment[];
  toUpdate: Assignment[];
}

export const getPayload = (
  selectedOrgs: string[],
  selectedPlan: PlanDefinition,
  selectedJurisdiction: OpenSRPJurisdiction,
  initialOrgs: string[] = []
): Assignment[] => {
  const now = moment(new Date());
  const planStart = moment(selectedPlan.effectivePeriod.start);
  const startDate = planStart > now ? now.format() : planStart.format();
  const endDate = moment(selectedPlan.effectivePeriod.end).format();
  const payload: Assignment[] = [];

  for (const orgId of selectedOrgs) {
    payload.push({
      fromDate: startDate,
      jurisdiction: selectedJurisdiction.id,
      organization: orgId,
      plan: selectedPlan.identifier,
      toDate: endDate,
    });
  }

  for (const retiredOrdId of initialOrgs.filter(orgId => !selectedOrgs.includes(orgId))) {
    payload.push({
      fromDate: startDate,
      jurisdiction: selectedJurisdiction.id,
      organization: retiredOrdId,
      plan: selectedPlan.identifier,
      toDate: moment(now.subtract(1, 'day')).format(),
    });
  }

  return payload;
};
