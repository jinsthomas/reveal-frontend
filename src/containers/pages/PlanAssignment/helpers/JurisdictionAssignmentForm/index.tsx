import { ErrorMessage, Field, Form, Formik, FormikActions } from 'formik';
import React, { Fragment } from 'react';
import { Button, Col, FormGroup, Row } from 'reactstrap';
import { SelectField, SelectOption } from '../../../../../components/TreeWalker/SelectField';
import { OpenSRPJurisdiction } from '../../../../../components/TreeWalker/types';
import {
  CLOSE,
  COULD_NOT_LOAD_FORM,
  DID_NOT_SAVE_SUCCESSFULLY,
  SAVE,
  SAVING,
  TEAM_ASSIGNMENT_SUCCESSFUL,
} from '../../../../../configs/lang';
import { PlanDefinition } from '../../../../../configs/settings';
import { OPENSRP_POST_ASSIGNMENTS_ENDPOINT } from '../../../../../constants';
import { successGrowl } from '../../../../../helpers/utils';
import { OpenSRPService } from '../../../../../services/opensrp';
import { Assignment } from '../../../../../store/ducks/opensrp/assignments';
import { getPayload } from './helpers';

export interface AssignmentFormProps {
  cancelCallBackFunc: () => void;
  defaultValue: SelectOption[];
  existingAssignments: Assignment[];
  jurisdiction: OpenSRPJurisdiction | null;
  messages: {
    fieldError: string;
    loadFormError: string;
  };
  options: SelectOption[];
  plan: PlanDefinition | null;
  submitCallBackFunc: (assignments: Assignment[]) => void;
  successNotifierBackFunc: (message: string) => void;
}

export const defaultAssignmentProps: AssignmentFormProps = {
  cancelCallBackFunc: () => null,
  defaultValue: [],
  existingAssignments: [],
  jurisdiction: null,
  messages: {
    fieldError: DID_NOT_SAVE_SUCCESSFULLY,
    loadFormError: COULD_NOT_LOAD_FORM,
  },
  options: [],
  plan: null,
  submitCallBackFunc: () => null,
  successNotifierBackFunc: successGrowl,
};

interface FormValues {
  organizations: string[];
}

const JurisdictionAssignmentForm = (props: AssignmentFormProps) => {
  const {
    cancelCallBackFunc,
    defaultValue,
    existingAssignments,
    jurisdiction,
    messages,
    options,
    plan,
    submitCallBackFunc,
    successNotifierBackFunc,
  } = props;

  if (!jurisdiction || !plan) {
    return <span>{messages.loadFormError}</span>;
  }

  const initialValues: FormValues = {
    organizations: defaultValue.map(e => e.value),
  };

  const submitForm = (
    values: FormValues,
    { setSubmitting, setFieldError }: FormikActions<FormValues>
  ) => {
    setSubmitting(true);
    const payload = getPayload(
      values.organizations,
      plan,
      jurisdiction,
      initialValues.organizations,
      existingAssignments
    );
    const OpenSrpAssignmentService = new OpenSRPService(OPENSRP_POST_ASSIGNMENTS_ENDPOINT);
    OpenSrpAssignmentService.create(payload)
      .then(response => {
        if (response) {
          successNotifierBackFunc(TEAM_ASSIGNMENT_SUCCESSFUL);
          submitCallBackFunc(payload);
        } else {
          setFieldError('organizations', messages.fieldError);
        }
      })
      .finally(() => setSubmitting(false))
      .catch((error: Error) => setFieldError('organizations', error.message));
  };

  return (
    <Fragment>
      <Formik initialValues={initialValues} onSubmit={submitForm}>
        {({ errors, isSubmitting }) => (
          <Form className="mb-5">
            <FormGroup className="select-orgs">
              <Field
                className={errors.organizations ? 'form-control is-invalid' : 'form-control'}
                component={SelectField}
                defaultValue={defaultValue}
                name="organizations"
                options={options}
              />
              <ErrorMessage
                name="organizations"
                component="p"
                className="form-text text-danger assignments-error"
              />
            </FormGroup>
            <FormGroup className="submit-group" id="submit-group">
              <Row>
                <Col md="6">
                  <Button
                    className="btn-block cancel"
                    size="sm"
                    type="button"
                    onClick={cancelCallBackFunc}
                  >
                    {CLOSE}
                  </Button>
                </Col>
                <Col md="6">
                  <Button
                    className="btn-block submit"
                    color="primary"
                    size="sm"
                    type="submit"
                    disabled={isSubmitting}
                  >
                    {isSubmitting ? SAVING : SAVE}
                  </Button>
                </Col>
              </Row>
            </FormGroup>
          </Form>
        )}
      </Formik>
    </Fragment>
  );
};

JurisdictionAssignmentForm.defaultProps = defaultAssignmentProps;

export { JurisdictionAssignmentForm };
