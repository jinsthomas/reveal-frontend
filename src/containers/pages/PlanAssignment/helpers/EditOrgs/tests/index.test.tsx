import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import flushPromises from 'flush-promises';
import React from 'react';
import { plans } from '../../../../../../store/ducks/opensrp/PlanDefinition/tests/fixtures';
import { defaultAssignmentProps } from '../../JurisdictionAssignmentForm';
import { assignment4, openSRPJurisdiction } from '../../JurisdictionAssignmentForm/tests/fixtures';
import { EditOrgs } from '../index';

describe('PlanAssignment/EditOrgs', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('works as expected', async () => {
    const submitMock: any = jest.fn();
    const growlMock: any = jest.fn();

    const props = {
      defaultValue: [{ label: 'Team X', value: 'x' }],
      existingAssignments: [assignment4],
      jurisdiction: openSRPJurisdiction,
      options: [
        { label: 'Team X', value: 'x' },
        { label: 'Team Y', value: 'y' },
      ],
      plan: plans[1],
      submitCallBackFunc: submitMock,
      successNotifierBackFunc: growlMock,
    };

    const wrapper = mount(<EditOrgs {...props} />);

    // initially there is just a button JurisdictionAssignmentForm
    expect(toJson(wrapper.find('button.show-form'))).toMatchSnapshot('Show form button');
    expect(wrapper.find('JurisdictionAssignmentForm').length).toEqual(0);

    // we can click the button to show the form
    (wrapper.find('button.show-form').props() as any).onClick();
    await flushPromises();
    wrapper.update();

    // now there is no show form button, but there is a form
    // note that there are other buttons, from the form
    expect(wrapper.find('button.show-form').length).toEqual(0);
    expect(wrapper.find('JurisdictionAssignmentForm').length).toEqual(1);
    expect(wrapper.find('JurisdictionAssignmentForm').props()).toEqual({
      ...props,
      assignTeamsLabel: 'Assign Teams',
      cancelCallBackFunc: expect.any(Function),
      labels: defaultAssignmentProps.labels,
    });

    // we can click the close button form to go back to before
    (wrapper.find('button.cancel').props() as any).onClick();
    await flushPromises();
    wrapper.update();

    // voila
    expect(wrapper.find('button.show-form').length).toEqual(1);
    expect(wrapper.find('JurisdictionAssignmentForm').length).toEqual(0);

    wrapper.unmount();
  });
});
