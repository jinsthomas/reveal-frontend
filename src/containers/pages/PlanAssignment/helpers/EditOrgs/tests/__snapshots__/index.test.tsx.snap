// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanAssignment/EditOrgs works as expected: Show form button 1`] = `
<button
  aria-label={null}
  className="show-form btn btn-primary btn-sm"
  onClick={[Function]}
  type="button"
>
  Assign Teams
</button>
`;
